### PR TITLE
Pass user authentication to Tiller

### PR DESCRIPTION
**Pull request by tamalsaha**
First step for #1918

### CODE REVIEW

**Comment by TamalSaha** [comment]
I signed it!

**Comment by TamalSaha** [comment]
@technosophos , here are few issues:

 - I don't see how to pass the updated context for StreamInterceptor.
 - KubeConfig has some extra options, like `impersonate`, `insecureSkipVerify`. https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/api/v1/types.go#L84 Does those need to be supported?

**Comment by technosophos** [comment]
I think we should probably disallow both `impersonate` and `insecureSkipVerify`.

**Comment by technosophos** [comment]
I put this onto 2.3.0 and marked it WIP so it doesn't actually get merged into 2.2.0

**Comment by saumanbiswas** [comment]
I signed it!

**Comment by TamalSaha** [comment]
@technosophos do you mind taking a first pass at this pr. If the general structure looks ok, we can write unit tests.

I have a follow up pr to add authZ check. https://github.com/appscode/helm/pull/10

**Comment by technosophos** [comment]
This will definitely need some documentation, but things are looking good. I am going to do some manual testing today.

**Comment by michelleN** [comment]
Just curious. How come the "Released By" field doesn't get populated by default with "admin" on a cluster with no users/roles (RBAC not set up)?

**Comment by TamalSaha** [comment]
@michelleN , what are you seeing in such a cluster? What type of of auth are you using (basic/bearer/client cert)?

**Comment by michelleN** [comment]
@TamalSaha it's a standard cluster on GKE

**Comment by michelleN** [comment]
@TamalSaha I'm using client cert

**Comment by TamalSaha** [comment]
Hi, any comments on the code?

File: pkg/helm/option.go
Comment (adamreese) [inline pkg/helm/option.go]: pkg/helm should not parse kubeconfig, it should be passed into the context from cmd/helm

Comment (adamreese) [inline pkg/helm/option.go]: there is a function that reads the config at https://github.com/kubernetes/helm/blob/master/cmd/helm/helm.go#L238-L250

File: pkg/helm/option.go
Comment (adamreese) [inline pkg/helm/option.go]: I don't think host is required, tiller already has it

File: pkg/helm/option.go
Comment (adamreese) [inline pkg/helm/option.go]: tiller already has the CA file

File: pkg/helm/option.go
Comment (technosophos) [inline pkg/helm/option.go]: Why not just clean `c` of fields we don't want, serialize it to JSON, and then send that instead of creating a whole bunch of different headers? That would simplify both the marshal and unmarshal operations for this data.

gRPC will even base64 encode/decode this for you if you suffix the header `-bin`. https://godoc.org/google.golang.org/grpc/metadata#New

File: pkg/helm/types.go
Comment (technosophos) [inline pkg/helm/types.go]: This file should be named `auth-header.go`, not `types.go`.

File: pkg/tiller/server.go
Comment (adamreese) [inline pkg/tiller/server.go]: i think this needs to be removed

File: pkg/tiller/server.go
Comment (adamreese) [inline pkg/tiller/server.go]: this needs to be checked before accessing to avoid panics

File: pkg/helm/types.go
Comment (technosophos) [inline pkg/helm/types.go]: If these are used by both the client and the server, they probably shouldn't be in the client package. That requires importing the client package for the server.

File: pkg/tiller/server.go
Comment (technosophos) [inline pkg/tiller/server.go]: This seems really complicated for what you are trying to do. Why not construct a single client config object and just run a `client.DiscoveryClient.ServerVersion()`? Wouldn't that just let the client library do most of what you do in the multiple functions below? This would be even easier if you sent a sanitized client config instead of storing various bits of the client config in different gRPC headers.

**Review by adamreese — CHANGES_REQUESTED** [review]

**Review by technosophos — CHANGES_REQUESTED** [review]
This PR seems to reproduce lots of stuff that's already in either Kubernetes or in Helm. Some of it appears to be copied verbatim from Kubernetes. It looks to me like it could be significantly simplified.

File: pkg/helm/types.go
Comment (adamreese) [inline pkg/helm/types.go]: `auth_header.go`

**Comment by TamalSaha** [comment]
@technosophos @adamreese PTAL. All comments should be addressed.

File: cmd/helm/installer/uninstall.go
Comment (bacongobbler) [inline cmd/helm/installer/uninstall.go]: this should be moved back. It's a local package.

Typical best practices for go imports are

1. stdlib
2. dependencies
3. local packages

`go fmt` would agree.

File: pkg/tiller/server.go
Comment (bacongobbler) [inline pkg/tiller/server.go]: same here RE: package imports

File: cmd/helm/installer/uninstall.go
Comment (TamalSaha) [inline cmd/helm/installer/uninstall.go]: ` gofmt -s -w cmd` does not say anything. But I can do that.

Comment (TamalSaha) [inline cmd/helm/installer/uninstall.go]: @bacongobbler , gofmt is ok with things as is.

Comment (adamreese) [inline cmd/helm/installer/uninstall.go]: @TamalSaha run `goimports` to group them correctly

Comment (TamalSaha) [inline cmd/helm/installer/uninstall.go]: @adamreese I just ran:
```
$ goimports -w cmd
$ goimports -w ./cmd/..
$ goimports -w ./pkg/..
$ goimports -w pkg
```
Still no change. In fact that make changes some other unrelated files:
```
	modified:   pkg/chartutil/load.go
	modified:   pkg/releaseutil/filter_test.go
	modified:   pkg/releaseutil/sorter_test.go
```
So, I skipped those.

Comment (bacongobbler) [inline cmd/helm/installer/uninstall.go]: Indeed, it seems like they enjoy it, still the precedence for packages being sorted in this order has always been implied. See [cmd/helm/init.go[(https://github.com/kubernetes/helm/blob/master/cmd/helm/init.go) (and anywhere else in the source code) for an example. It's just best practices we've been following.

Comment (TamalSaha) [inline cmd/helm/installer/uninstall.go]: Never heard of this best practice. But if project follows unwritten best practices regarding style, I'm happy to oblige. Just thought I was done with these, when I switched to writing GO. :)

File: cmd/helm/installer/uninstall.go
Comment (bacongobbler) [inline cmd/helm/installer/uninstall.go]: this got accidentally shifted over

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: cmd/helm/installer/uninstall.go
Comment (TamalSaha) [inline cmd/helm/installer/uninstall.go]: fixed.

**Comment by TamalSaha** [comment]
I have added the checks for subject review in this pr. Since each request is authenticated using caller info, `KubeClient` can't be initiated when the server starts. KubeClient is now created inside the interceptor and stored in context for `ReleaseServer` to use it.

**Comment by technosophos** [comment]
@TamalSaha Sorry, I lost track of this PR last week. Is it ready for a new round of reviews?

**Comment by TamalSaha** [comment]
@technosophos yes!

File: cmd/helm/helm.go
Comment (technosophos) [inline cmd/helm/helm.go]: panicing will cause a stack trace to be displayed to the end user. Do we really want that?

**Comment by technosophos** [comment]
Can you please break out the subject review stuff into a separate PR as we originally agreed? https://github.com/kubernetes/helm/issues/1918#issuecomment-278191152

It's just too many moving parts for one PR -- especially one we were already in the midst of reviewing.

**Comment by TamalSaha** [comment]
@technosophos I have removed the subreview related changes. Also addressed your other comment about panic. PTAL.

**Comment by TamalSaha** [comment]
@technosophos , I have brought back adding username to release. I mistakenly removed it when reverting subject review checks.

**Comment by technosophos** [comment]
@TamalSaha this needs a rebase to fix CircleCI

**Comment by TamalSaha** [comment]
@technosophos rebased and tests are passing.

**Comment by TamalSaha** [comment]
FYi, I believe, https://github.com/kubernetes/helm/issues/2172 needs to be resolved too.

**Comment by technosophos** [comment]
I am sorry for not getting this reviewed in time for 2.3. That was my fault. I think we can take a look at getting #2172 resolved as soon as we switch to the Kube 1.6 libraries in the early 2.4 cycle. So that should get us going on this again.

**Comment by TamalSaha** [comment]
I have squashed the commits so that it is easy to rebase.

**Comment by TamalSaha** [comment]
@michelleN , it seems that GKE does not support TokenReview api 9even with alpha features enabled in 1.5.6 cluster).

**Comment by TamalSaha** [comment]
@adamreese / @technosophos , to test this manually, deploy this tiller. Then install a new chart usign helm build from this pr. Then list releases.  You should see the deployer name.

```
$ helm install stable/redis
$ helm ls

# ./helm install stable/redis
NAME:   factual-lionfish
LAST DEPLOYED: Mon Apr 10 16:31:03 2017
NAMESPACE: default
STATUS: DEPLOYED
RELEASED BY: qacode.system-admin

RESOURCES:
==> v1/Secret
NAME                    TYPE    DATA  AGE
factual-lionfish-redis  Opaque  1     1s

==> v1/PersistentVolumeClaim
NAME                    STATUS   VOLUME  CAPACITY  ACCESSMODES  AGE
factual-lionfish-redis  Pending  1s

==> v1/Service
NAME                    CLUSTER-IP   EXTERNAL-IP  PORT(S)   AGE
factual-lionfish-redis  10.0.194.85  <none>       6379/TCP  1s

==> extensions/v1beta1/Deployment
NAME                    DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
factual-lionfish-redis  1        1        1           0          1s


NOTES:
Redis can be accessed via port 3306 on the following DNS name from within your cluster:
factual-lionfish-redis.default.svc.cluster.local

To get your password run:

    REDIS_PASSWORD=$(kubectl get secret --namespace default factual-lionfish-redis -o jsonpath="{.data.redis-password}" | base64 --decode)

To connect to your Redis server:

1. Run a Redis pod that you can use as a client:

   kubectl run factual-lionfish-redis-client --rm --tty -i --env REDIS_PASSWORD=$REDIS_PASSWORD --image bitnami/redis:3.2.8-r2 -- bash

2. Connect using the Redis CLI:

  redis-cli -h factual-lionfish-redis -a $REDIS_PASSWORD

root@qa-buildbox:~# ./helm ls
NAME               	REVISION	UPDATED                 	STATUS  	CHART      	NAMESPACE	RELEASED BY        
factual-lionfish   	1       	Mon Apr 10 16:31:03 2017	DEPLOYED	redis-0.5.1	default  	qacode.system-admin
fancy-chinchilla   	1       	Mon Apr 10 08:02:27 2017	DEPLOYED	redis-0.5.1	default  	                   
kissable-woodpecker	1       	Mon Apr 10 07:59:31 2017	DEPLOYED	mysql-0.2.5	default  	                   
```

**Comment by TamalSaha** [comment]
I am told that GKE does not enable TokenReview in sig-apps. So, once auth is added Tiller will not work in GKE?

sig-app-tiller

**Comment by TamalSaha** [comment]
I have added checks to dynamically check if TokenReview api is not enabled. I don't get the error in GKE any more.

**Comment by technosophos** [comment]
This should be unblocked again at this point. We've merged in gRPC 1.2.

**Comment by technosophos** [comment]
Bumping to 2.5.0, since it still needs rebasing.

**Comment by technosophos** [comment]
Refactor is done. This issue is unblocked.

**Comment by mattfarina** [comment]
Any movement on this happening?

**Comment by TamalSaha** [comment]
Rebased.

**Comment by seh** [comment]
Do you intend to take group membership into account as well? I see treatment of "user names" in the proposed changes [here in file _pkg/tiller/server.go_](https://github.com/appscode/helm/blob/586a496fca64f6aeac84e0a64a52900c74e0d705/pkg/tiller/server.go#L268-L270) we're omitting the groups that you can extract from the "Organization" AVAs in the certificate's subject.

File: pkg/tiller/server.go
Comment (seh) [inline pkg/tiller/server.go]: - [ ] What about using [`crt.Subject.Organization`](https://golang.org/pkg/crypto/x509/pkix/#Name) here to populate the `UserInfo` "Groups" field?

Comment (TamalSaha) [inline pkg/tiller/server.go]: Group name needs to be preserved for TokenReview calls. But I believe this was not next working at Kube 1.5, when I originally wrote this pr.

Comment (seh) [inline pkg/tiller/server.go]: The authentication documentation on X.509 client certificates is [here](https://kubernetes.io/docs/admin/authentication/#x509-client-certs).
> As of Kubernetes 1.4, client certificates can also indicate a user’s group memberships using the certificate’s organization fields. To include multiple group memberships for a user, include multiple organization fields in the certificate.

In my clusters, we write most of our RBAC bindings against groups, not users, so honoring the group membership expressed in the certificates' "O" AVAs is important for us.

**Comment by tamalsaha** [comment]
@technosophos, it seems that I won't be able to find time to work through the review process. I am sad that it turned out this way. I don't want to hold off this feature from Helm users. So, I am closing this pr. Core team / interested parties are welcome to work through the review process based on this pr.

**Pull request closed without merging**
